Cython/Compiler/ExprNodes.py

```diff
@@ -500,18 +500,23 @@ def result(self):
         else:
             return self.calculate_result_code()
 
+    def _make_move_result(self, result):
+        self.has_temp_moved = True
+        return "__PYX_STD_MOVE_IF_SUPPORTED({0})".format(result)
+
     def _make_move_result_rhs(self, result, allow_move=True):
         if (self.is_temp and allow_move and
                 self.type.is_cpp_class and not self.type.is_reference):
-            self.has_temp_moved = True
-            return "__PYX_STD_MOVE_IF_SUPPORTED({0})".format(result)
+            return self._make_move_result(result)
         else:
             return result
 
     def move_result_rhs(self):
         return self._make_move_result_rhs(self.result())
 
     def move_result_rhs_as(self, type):
+        if type.is_rvalue_reference:
+            return self._make_move_result(self.result_as(type))
```

Comment:
I wonder if this should generate `std::move` rather than `__PYX_MOVE_IF_POSSIBLE`. `__PYX_MOVE_IF_POSSIBLE` was designed as an improvement for C++11 that didn't break C++98 code?

If the type is an r-value reference then it's definitely going to require C++11 and so there's no need to generate the "compatible" version (which won't work here anyway).

Reply:
I changed to use `std::move`.